src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,295 @@
+
+#include "sync_permission.hpp"
+#include "sync_user.hpp"
+#include "sync_config.hpp"
+#include "sync_manager.hpp"
+#include "sync_session.hpp"
+#include "object_schema.hpp"
+#include "impl/object_accessor_impl.hpp"
+
+#if REALM_PLATFORM_APPLE
+#include <realm/util/cf_ptr.hpp>
+#else
+#include <sole.hpp>
+#endif
+
+#include "property.hpp"
+
+using namespace realm;
+
+namespace {
+
+// TODO: this should already exist somewhere.
+struct MallocBuffer {
+    void* buffer_ptr;
+
+    MallocBuffer(size_t bytes)
+    : buffer_ptr(malloc(bytes))
+    {
+        if (!buffer_ptr)
+            throw std::runtime_error("Could not successfully malloc memory.");
+    }
+
+    ~MallocBuffer()
+    {
+        free(buffer_ptr);
+    }
+
+    template<typename T>
+    T* get()
+    {
+        return static_cast<T*>(buffer_ptr);
+    }
+};
+
+std::string make_uuid()
+{
+#if REALM_PLATFORM_APPLE
+    CFPtr<CFUUIDRef> uuid = adoptCF(CFUUIDCreate(NULL));
+    CFPtr<CFStringRef> uuid_str = adoptCF(CFUUIDCreateString(NULL, uuid.get()));
+    size_t buffer_len = CFStringGetLength(uuid_str.get()) + 1;
+    MallocBuffer buffer(buffer_len);
+    CFStringGetCString(uuid_str.get(), buffer.get<char>(), buffer_len, kCFStringEncodingASCII);
+    std::string uuid_cpp_str = buffer.get<char>();
+    return uuid_cpp_str;
+#else
+    return sole::uuid4().str();
+#endif
+}
+
+}
+
+size_t PermissionResults::size() {
+    REALM_ASSERT(m_results->size());
+    return m_results->size() - 1;
+}
+
+#pragma mark - Permission
+
+Permission::Permission(const Permission& p)
+: path(p.path)
+, access(p.access)
+, condition(p.condition)
+{ }
+
+Permission::Permission(std::string path, AccessLevel access, Condition condition)
+: path(std::move(path))
+, access(std::move(access))
+, condition(std::move(condition))
+{ }
+
+Permission& Permission::operator=(const Permission& p)
+{
+    if (&p != this) {
+        path = p.path;
+        access = p.access;
+        condition = p.condition;
+    }
+    return *this;
+}
+
+Permission::Condition& Permission::Condition::operator=(const Permission::Condition& c)
+{
+    if (&c != this) {
+        type = c.type;
+        switch (type) {
+            case Type::UserId:
+                user_id = c.user_id;
+                break;
+            case Type::KeyValue:
+                key_value = c.key_value;
+                break;
+        }
+    }
+    return *this;
+}
+
+Permission::Condition::Condition(const Permission::Condition& c)
+: type(c.type)
+{
+    switch (type) {
+        case Type::UserId:
+            user_id = c.user_id;
+            break;
+        case Type::KeyValue:
+            key_value = c.key_value;
+            break;
+    }
+}
+
+Permission::Condition::~Condition()
+{
+    switch (type) {
+        case Type::UserId:
+            user_id.std::basic_string<char>::~basic_string<char>();
+            break;
+        case Type::KeyValue:
+            std::get<0>(key_value).std::basic_string<char>::~basic_string<char>();
```

Comment:
This should invoke `key_value`'s destructor rather than manually invoking the destructor on each of its members.